src/natcap/invest/spec.py

```diff
@@ -0,0 +1,2042 @@
+            'MULTILINESTRING': [ogr.wkbMultiLineString, ogr.wkbMultiLineStringM,
+                                ogr.wkbMultiLineStringZM,
+                                ogr.wkbMultiLineString25D],
+            'MULTIPOLYGON': [ogr.wkbMultiPolygon, ogr.wkbMultiPolygonM,
+                             ogr.wkbMultiPolygonZM, ogr.wkbMultiPolygon25D]
+        }
+
+        allowed_geom_types = []
+        for geom in self.geometries:
+            allowed_geom_types += geom_map[geom]
+
+        # NOTE: this only checks the layer geometry type, not the types of the
+        # actual geometries (layer.GetGeometryTypes()). This is probably equivalent
+        # in most cases, and it's more efficient than checking every geometry, but
+        # we might need to change this in the future if it becomes a problem.
+        # Currently not supporting ogr.wkbUnknown which allows mixed types.
+        layer = gdal_dataset.GetLayer()
+        if layer.GetGeomType() not in allowed_geom_types:
+            return get_message('WRONG_GEOM_TYPE').format(allowed=self.geometries)
+
+        if self.fields:
+            field_patterns = []
+            for spec in self.fields:
+                # brackets are a special character for our args spec syntax
+                # they surround the part of the key that's user-defined
+                # user-defined rows/columns/fields are not validated here, so skip
+                if spec.required is True and '[' not in spec.id:
+                    field_patterns.append(spec.id)
+
+            fieldnames = [defn.GetName() for defn in layer.schema]
+            required_field_warning = check_headers(
+                field_patterns, fieldnames, 'field')
+            if required_field_warning:
+                return required_field_warning
+
+        srs = layer.GetSpatialRef()
+        projection_warning = _check_projection(srs, self.projected, self.projection_units)
+        return projection_warning
+
+
+@dataclasses.dataclass
+class RasterOrVectorInput(SingleBandRasterInput, VectorInput):
+    """An invest model input that can be either a raster or a vector.
+
+    Attributes:
+        geometries: A set of geometry type(s) that are allowed for this vector
+        fields: An iterable of `Input`s representing the fields that this
+            vector is expected to have. The `key` of each input must match the
+            corresponding field name.
+        projected: Defaults to None, indicating a projected (as opposed to
+            geographic) coordinate system is not required. Set to True if a
+            projected coordinate system is required.
+        projection_units: Defaults to None. If `projected` is `True`, and a
+            specific unit of projection (such as meters) is required, indicate
+            it here.
+    """
+    band: typing.Union[Input, None] = None
+    geometries: set = dataclasses.field(default_factory=dict)
+    fields: typing.Union[typing.Iterable[Input], None] = None
```

Comment:
Do we need to define these here since they're already defined in the parent classes of `SingleBandRasterInput` or `VectorInput`?

Reply:
You're right, they do not need to be redefined here!